Improved Looper realism

kriegfrj opened the pull request:
Apart from some general tidy-ups, this PR does a few things:

* The biggest change came about because I noticed during my testing & auditing that there is no mechanism to close background loopers at the end of a test. I put some code into `RobolectricTestRunner` to dump the active threads at the end of each test, and I found that at the end of a full test suite run there were dozens of looper threads in `Object.wait()`. These threads were never exiting and hence never releasing their associated loopers and their resources. I suspect that this is the real source of the memory leak that the `SoftThreadLocal` in `ShadowLooper` was attempting to solve. This PR fixes `ShadowLooper.resetThreadLoopers()` to make sure that `quit()` is called on all background loopers. By the end of the full test suite they've all exited.
* Facilitating the above, I replaced the `SoftThreadLocal` with a `WeakHashMap` so that it is possible to see what a thread's looper is from another thread. This also allowed me to implement a utility function that will give you the looper for a given thread. It is also now possible (though yet implemented) for one thread to initialise another thread's looper, which might simplify some test scenarios.
* Having solved the memory leak issue (fingers crossed!) I was emboldened to remove some of the shadowing and use more real Android. In particular, I noted that our shadow implementation of `myLooper()` would implicitly do a `prepare()` for the thread too, whereas in Android calling `myLooper()` without first calling `prepare()` would result in a runtime exception. However fixing this issue also made me run into #1407 with `ShadowValueAnimatorTest` which I had to work around (I'll comment more on what I discovered against that issue).
* There are no longer any references to `ShadowApplication` from within `ShadowLooper`, making `ShadowLooper` better encapsulated & decoupled from `ShadowApplication`. This has no immediate effect but was a prerequisite to my attempt to implement a master scheduler - I was running into a chicken-and-egg problem when trying to initialise the main looper and its main scheduler. This also means that you can call `getMainLooper()` before the `ShadowApplication` is set up and while `RuntimeEnvironment.application` is still null, and you'll still get the main looper (used to throw an NPE in this case, which meant that background threads would occasionally be stillborn).

jongerrish commented on robolectric/src/test/java/org/robolectric/shadows/ShadowLooperTest.java:
```diff
@@ -5,28 +5,106 @@
+      super(testName.getMethodName());
+    }
+    
+//    public QuitThread(String name) {
```
Remove commented out code?

jongerrish commented on robolectric/src/test/java/org/robolectric/shadows/ShadowLooperTest.java:
```diff
@@ -116,15 +169,39 @@ public void run() {
+    assertThat(test.hasContinued).as("afterJoin").isTrue();
+  }
+
+//  @Test
```
Why is the test commented out?

jongerrish commented:
This looks great! Is it likely to break many tests? Would you like me to patch and run through our submit queue?

jongerrish commented:
Doesn't look like this change breaks any of our tests :-) I'd be happy to have this merged @erd ?

mag commented:
I haven't had time to absorb the code in this PR, so apologies in advance, but let me ask a dumb question: under what situations are loopers running on threads other than the main thread? I frequently advise people that Robolectric should only be used in a single-threaded context, and I was pretty sure that most normal/supported use cases do not launch threads. @kriegfrj, why are your tests launching background loopers? Does this PR mean that we are explicitly supporting (and advising) multi-threaded tests?

jongerrish commented:
I don't think we are advocating multi threaded tests, I would still
advocate unit testing a single thread of control only.

The reason there is code to handle multiple threads is because android
itself creates threads, so does Robolectric internally in some cases where
needed, e.g: SqliteConnections must be accessed on a separate thread (and
we take care of the synchronization).

We would still advocate single threaded testing though.
On Jul 9, 2015 12:02 PM, "Mike Grafton" <notifications@github.com> wrote:

> I haven't had time to absorb the code in this PR, so apologies in advance,
> but let me ask a dumb question: under what situations are loopers running
> on threads other than the main thread? I frequently advise people that
> Robolectric should only be used in a single-threaded context, and I was
> pretty sure that most normal/supported use cases do not launch threads.
> @kriegfrj <https://github.com/kriegfrj>, why are your tests launching
> background loopers? Does this PR mean that we are explicitly supporting
> (and advising) multi-threaded tests?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/robolectric/robolectric/pull/1931#issuecomment-120109157>
> .
>

mag commented:
@jongerrish Sweet. Thanks for the clarification.

kriegfrj commented:
Thanks for the comments @jongerrish. I will have an updated PR shortly addressing the commented out code.

I see that you have already found out for yourself, but I anticipate that these changes shouldn't break any tests unless you are using timeouts (per the comment in the OP) or else in some other way relying on `myLooper()` to work without a prior call to `prepare()`. I'm not too concerned about this because that would probably be a bug anyway, and if it isn't it's easy to work around by slipping in a call to `prepare()` at the appropriate place.

Regarding the policy on multi-threaded testing: I think with these changes, in conjunction with my next PR (implementing point 5 of #1879), we might be able to support a degree of multi-threaded testing. Specifically, we should be able to support multi-threaded testing as long as all the threads you create are `Looper` threads. To explain: the reason that multi-threaded testing is discouraged for testing (at least, as I understand it) is that you can't control execution order in multi-threaded code. This makes the test execution order unpredictable, which leads to flakiness, which leads to a loss of confidence in the tests + time wasted in debugging the associated issues, etc, etc. With the upcoming PR, although there are multiple looper threads, we shouldn't have these problems because there will still be a single thread of control (the main test thread) - execution of all threads will be serialised (only one thread executing at a time) and the order & timing in which they execute will be controllable by the test thread thus making it possible to achieve test execution which is 100% repeatable and eliminating flakiness. At least, that's the goal.

erd commented:
> I'd be happy to have this merged @erd ?

You guys are creating pull-requests faster than I can absorb them :smile:  Not that I'm complaining...
